// @redocly/openapi-cli/lib/commands/split/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OPENAPI3_COMPONENT_NAMES = exports.OPENAPI3_COMPONENT = exports.OPENAPI3_METHOD_NAMES = exports.componentsPath = exports.PATHS = exports.COMPONENTS = void 0;
exports.COMPONENTS = 'components';
exports.PATHS = 'paths';
exports.componentsPath = `#/${exports.COMPONENTS}/`;
var OPENAPI3_METHOD;
(function (OPENAPI3_METHOD) {
    OPENAPI3_METHOD["Get"] = "get";
    OPENAPI3_METHOD["Put"] = "put";
    OPENAPI3_METHOD["Post"] = "post";
    OPENAPI3_METHOD["Delete"] = "delete";
    OPENAPI3_METHOD["Options"] = "options";
    OPENAPI3_METHOD["Head"] = "head";
    OPENAPI3_METHOD["Patch"] = "patch";
    OPENAPI3_METHOD["Trace"] = "trace";
})(OPENAPI3_METHOD || (OPENAPI3_METHOD = {}));
exports.OPENAPI3_METHOD_NAMES = [
    OPENAPI3_METHOD.Get,
    OPENAPI3_METHOD.Put,
    OPENAPI3_METHOD.Post,
    OPENAPI3_METHOD.Delete,
    OPENAPI3_METHOD.Options,
    OPENAPI3_METHOD.Head,
    OPENAPI3_METHOD.Patch,
    OPENAPI3_METHOD.Trace
];
var OPENAPI3_COMPONENT;
(function (OPENAPI3_COMPONENT) {
    OPENAPI3_COMPONENT["Schemas"] = "schemas";
    OPENAPI3_COMPONENT["Responses"] = "responses";
    OPENAPI3_COMPONENT["Parameters"] = "parameters";
    OPENAPI3_COMPONENT["Examples"] = "examples";
    OPENAPI3_COMPONENT["Headers"] = "headers";
    OPENAPI3_COMPONENT["RequestBodies"] = "requestBodies";
    OPENAPI3_COMPONENT["Links"] = "links";
    OPENAPI3_COMPONENT["Callbacks"] = "callbacks";
    OPENAPI3_COMPONENT["SecuritySchemes"] = "securitySchemes";
})(OPENAPI3_COMPONENT = exports.OPENAPI3_COMPONENT || (exports.OPENAPI3_COMPONENT = {}));
exports.OPENAPI3_COMPONENT_NAMES = [
    OPENAPI3_COMPONENT.RequestBodies,
    OPENAPI3_COMPONENT.Schemas,
    OPENAPI3_COMPONENT.Responses,
    OPENAPI3_COMPONENT.Parameters,
    OPENAPI3_COMPONENT.Examples,
    OPENAPI3_COMPONENT.Headers,
    OPENAPI3_COMPONENT.Links,
    OPENAPI3_COMPONENT.Callbacks,
    OPENAPI3_COMPONENT.SecuritySchemes
];
